Ff150 `<img>` attribute `sizes` takes value `auto` (#43684)
mdn/content: files/en-us/web/html/reference/elements/img/index.md

Reviewer: ```suggestion
      - : Loads the image immediately, regardless of whether or not the image is currently within the {{glossary("visual viewport")}} (this is the default value).
```

Author: This is good, but github is showing it to me as an addition to an empty line (i.e. some kind of github bug). Setting resolved but I will make this change manually.

    - `unsafe-url`: The referrer will include the origin _and_ the path (but not the [fragment](/en-US/docs/Web/API/HTMLAnchorElement/hash), [password](/en-US/docs/Web/API/HTMLAnchorElement/password), or [username](/en-US/docs/Web/API/HTMLAnchorElement/username)). **This value is unsafe**, because it leaks origins and paths from TLS-protected resources to insecure origins.

- `sizes`
  - : One or more values separated by commas, which can be source sizes or the `auto` keyword.
    The spec requires that the `sizes` attribute only be present when `srcset` uses width descriptors.
    - **source size**
      - : A **source size** consists of:
        1. A [media condition](/en-US/docs/Web/CSS/Guides/Media_queries/Using#syntax), omitted for the last item in the list.
        2. A source size value.

>         For example, the following source size proposes using a `1000px`-wide image source if the _viewport_ height is 500px or less.

Author: ```suggestion
        For example, the following source size proposes using a `1000px`-wide image source if the _viewport_ width is 500px or less.
```

>         For example, the following source size proposes using an image source of 1000px width if the _viewport_ height is 500px or less.

Reviewer 2: ```suggestion
        For example, the following source size proposes using a `1000px`-wide image source if the _viewport_ height is 500px or less.
```

        ```css

>         (height <= 500px) 1000px

Author: ```suggestion
        (width<= 500px) 1000px
```

>         ```

Author: This section is the same information put into a definition list. The only change in structure otherwise is pulling out this example of what a query looks like.

>         Media conditions describe properties of the _viewport_, not the _image_.

Reviewer: ```suggestion
        Media conditions describe properties of the _{{glossary("viewport")}}_, not the _image_.
```

>         Because a source size descriptor specifies the width to use for the image during layout, the media condition is typically (but not necessarily) based on the {{cssxref("@media/width")}}.

Reviewer 2: In which case, it seems weird to show a `height` viewport example earlier on. I mean, you are nearly never going to want to do that.

Author: Yeah, that predates my changes.  MOdified the query to condition on the width (but did not change the width) in https://github.com/mdn/content/pull/43684/changes/59f86f56d36a3ddf134e32172cbad50e75c50fe9

Reviewer 2: Perfect, this makes a lot more sense now. I edited the file to add a space to make the code a bit neater.

        Source size values specify the intended display size of the image.
        {{glossary("User agent", "User agents")}} use the current source size to select one of the sources supplied by the `srcset` attribute, when those sources are described using width (`w`) descriptors.

>         The selected source size affects the {{glossary("intrinsic size")}} of the image (the image's display size if no {{glossary("CSS")}} styling is applied).

Reviewer 2: "affects" seems like a weird word to use here — you can't change an image's intrinsic size. I'm assuming what this really means is that it specifies the intrinsic size of the image that should be chosen.

So maybe something like this:

```suggestion
        The selected source size specifies the {{glossary("intrinsic size")}} of the image that should be chosen (the image's display size if no {{glossary("CSS")}} styling is applied).
```

?

Author: No, this actually means what it says, though it's a bit confusing. Generally we think about the intrinsic size as the raw pixel dimensions. But when using a srcset the width in the `sizes` takes this role. So if you specify a w value of 500 px that many pixels are reserved, and whatever image you download is scaled to fit into that space.

Perhaps we can avoid overloading the term intrinsic size - better?

```suggestion
        The `w` value defined in sizes determines the image’s default layout width.
        In the absence of {{glossary("CSS")}}, the browser will render the image at this size, regardless of the physical pixel dimensions of the downloaded file.
```

Reviewer 2: This is much better, thanks! I'm going to commit this suggestion.

(I've not looked at this stuff for ages, and had forgotten how confusing it is)

    A source size value can be any non-negative [length](/en-US/docs/Web/CSS/Reference/Values/length). It must not use CSS functions other than the [math functions](/en-US/docs/Web/CSS/Reference/Values/Functions#math_functions). Units are interpreted in the same way as [media queries](/en-US/docs/Web/CSS/Guides/Media_queries), meaning that all relative length units are relative to the document root rather than the `<img>` element. For example, an `em` value is relative to the root font size, not the font size of the image. [Percentage](/en-US/docs/Web/CSS/Reference/Values/percentage) values are not allowed. If the `sizes` attribute is not provided, it has a default value of `100vw` (the viewport width).
        Source size values specify the intended display size of the image.
        {{glossary("User agent", "User agents")}} use the current source size to select one of the sources supplied by the `srcset` attribute, when those sources are described using width (`w`) descriptors.
        The selected source size affects the {{glossary("intrinsic size")}} of the image (the image's display size if no {{glossary("CSS")}} styling is applied).

>     The `auto` keyword can replace the whole list of sizes or the first entry in the list. It is only valid when combined with `loading="lazy"`, and resolves to the [concrete size](/en-US/docs/Web/CSS/Reference/Values/image) of the image. Since the intrinsic size of the image is not yet known, `width` and `height` attributes (or CSS equivalents) should also be specified to prevent the browser from assuming the default image width of 300px.

Author: FYI I didn't think this was quite right. The browser might now know intrinsic sizes at some points but it doesn't need to - it falls back to the default intrinsic size for images in the ua stylesheet. Also this is not a factor of "auto" so much as the fallback when auto can't work. Such as if you have a lazy loaded image but it is displayed in the viewport on load (so it isn't lazy loaded)

>         That is, it should use the [concrete size](/en-US/docs/Web/CSS/Reference/Values/image) of the image, calculated after layout from HTML and CSS has been applied.

Reviewer: Maybe link to the description section of the `<image>` value, where "concrete size" is defined.

```suggestion
        That is, it should use the [concrete size](/en-US/docs/Web/CSS/Reference/Values/image#concrete_size) of the image, calculated after layout from HTML and CSS has been applied.
```

Author: It's a good idea. I added an explicit heading as a link target and modified your suggestion ^^^. Thank you.

>         This is only valid when combined with `loading="lazy"`, as the page can be expected to already have CSS and other layout information by the time the image is loaded.

Reviewer 2: ```suggestion
        This is only valid when combined with `loading="lazy"`, as the page is expected to already have CSS and other layout information by the time the image loads.
```

        Using `auto` saves you having to specify your layout media conditions twice: once for layout, and once for selection of an appropriate image to fetch and display.

>         If `auto` cannot resolve — either because the browser does not support it, or because the image has no layout size yet — the browser falls through to the fallback _source sizes_ in the list to determine the width, then to `width`/`height` attributes defined on the element, and finally to the default intrinsic size for `<img>` elements defined in the user agent stylesheet (300px by 150px).

Reviewer 2: ```suggestion
        If `auto` cannot resolve — either because the browser does not support it, or because the image has no layout size yet — the browser falls back to the _source sizes_ in the list to determine the width, then to `width`/`height` attributes defined on the element, and finally to the default intrinsic size for `<img>` elements defined in the user agent stylesheet (300px by 150px).
```
